### PR TITLE
feat(behavior_path_planner): abort lane change function

### PR DESCRIPTION
**Pull request by zulfaqar-azmi-t4**
## Description

Current lane change abort function only cancels unsafe lane change path when ego is still relatively close to the current reference lane (preparation phase). While ego is lane changing, there is no action to respond to the unsafe lane change path. Therefore the aim of this PR is to introduce a function that allows ego vehicle to return to original lane should abort is needed.

Note that this function "DOESN'T GUARANTEE SAFE ABORT"

abort-illustration

The parameter `abort_begin_dist` and `abort_return_dist` are computed as follows
```
dist = current_vel * time + 0.5 * accel * time ^ 2 - (1/6) * jerk * time ^ 3
```

are subjected to the following constraints

```
abort_begin_min_longitudinal_thresh < abort_begin_dist < abort_begin_max_longitudinal_thresh
abort_return_min_longitudinal_thresh < abort_return_dist < abort_return_max_longitudinal_thresh
```

### Flowchart

flowchart-abort-process

The abort process may result in two different outcome. If ego as yet no depart from the reference lane, the result is `CANCEL` process. (Note that this is the original abort function that has already available in the `main` branch)

The following illustrates the `CANCEL` process

https://user-images.githubusercontent.com/93502286/204445547-4518113a-00b1-4040-ba19-59679755e316.mp4

If ego depart from the current lane, then the result will be `Abort` process.

https://user-images.githubusercontent.com/93502286/204446006-2a45edf8-f06b-4093-b4ad-a1741eb420f8.mp4

## Related links

Parameters: https://github.com/tier4/autoware_launch/pull/583
[TIER4 Internal Link
](https://tier4.atlassian.net/browse/T4PB-21434)

## Tests performed

https://user-images.githubusercontent.com/93502286/203713966-6a5e8030-942b-400e-95cb-490f240e81f7.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by TakaHoribe — COMMENTED** [review]
@zulfaqar-azmi-t4 Thank you for this PR. Here is some comments:

- I don't think it's possible to pre-define the deceleration/jerk in the abort distance calculation. Often, when humans abort a lane change, they don't decelerate. It might even be dangerous to decelerate when aborting. If you have a clear policy in place, we can provide parameters, but it may not be necessary and 0 for acc/jerk is used instead in this case.

- The calculation of the abort path should not be based on distance, but should be calculated from lateral acceleration and lateral jerk. In this case, we can simply calculate the required distance as d = v * dt for now to reduce parameters.

- It's not essential to use an angle and distance to distinguish between abort and cancel functions. Since the difference between these functions is the necessity of the path generation, we can just check whether the "shift start point for lane change" is behind the "shift start point for abort". I think this is simpler and no threshold parameters are required. What do you think?

File: planning/behavior_path_planner/src/scene_module/lane_change/lane_change_module.cpp [hunk 1]
```diff
@@ -429,86 +475,85 @@ bool LaneChangeModule::isCurrentSpeedLow() const
...
+      status_.current_lanes, getEgoPose(), common_parameters);
+
+    if (is_within_original_lane) {
       return true;
     }
 
-    const bool is_distance_small =
-      lane_change_utils::isEgoDistanceNearToCenterline(closest_lanelet, current_pose, *parameters_);
+    // check abort enable flag
+    auto clock{rclcpp::Clock{RCL_ROS_TIME}};
```
[the remark this answers is not in the document]
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/lane_change/lane_change_module.cpp]: The clock is fixed in [392645c](https://github.com/autowarefoundation/autoware.universe/pull/2359/commits/392645c78d944960923fb400897267ce3290ca77)

File: planning/behavior_path_planner/config/lane_change/lane_change.param.yaml [hunk 2]
```diff
@@ -9,12 +9,37 @@
...
+      prepare_phase_ignore_target_speed_thresh: 0.1 # [m/s]
       use_predicted_path_outside_lanelet: true
       use_all_predicted_path: true
+
+      # abort
+      enable_cancel_lane_change: true
+      enable_abort_lane_change: true
+
+      abort_lane_change_velocity_thresh: 0.5
+      abort_lane_change_angle_thresh: 10.0      # [deg]
```
[the remark this answers is not in the document]
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/config/lane_change/lane_change.param.yaml:24]: the parameters are removed in [bf23a15](https://github.com/autowarefoundation/autoware.universe/pull/2359/commits/bf23a15504212b1633ebd89e44b56a965af48820)

File: planning/behavior_path_planner/src/scene_module/lane_change/util.cpp [hunk 3]
```diff
@@ -739,4 +724,172 @@ std::vector<DrivableLanes> generateDrivableLanes(
...
+        break;
+      }
+    }
+    return idx;
+  };
+
+  const auto abort_delta_time = lane_change_param.abort_delta_time;
+
+  double abort_start_dist{0.0};
+  const auto abort_start_idx = pose_idx_min(abort_delta_time, abort_start_dist);
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/lane_change/util.cpp:772]: how about make the output of `pose_idx_min` std::pair and return both `abort_start_dist` and `abort_start_idx`?

```
const auto pose_idx_min = [&](const double param_time)
{
   ...
   return std::pair{idx, turning_point_dist};
};
```

Then you can receive two outputs like this:
```
const auto [abort_start_dist, abort_start_idx] = pose_idx_min(abort_delta_time);
```

**Comment by zulfaqar-azmi-t4** [comment]
Evaluator pass with 820/825
[tier4 internal link to evaluator](https://evaluation.tier4.jp/evaluation/reports/9fa88c2f-80c4-5735-b906-4021b649c4e2?project_id=prd_jt)

File: planning/behavior_path_planner/src/scene_module/lane_change/util.cpp (hunk 3 above)
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/lane_change/util.cpp]: this was fixed in [4b61a8f](https://github.com/autowarefoundation/autoware.universe/pull/2359/commits/4b61a8f806b49cc4881b48255cbc89e6be65ac66)

**Review by TakaHoribe — APPROVED** [review]

**Pull request merged**